File: shiftzeroes/shiftzeroes.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Linq;

namespace ShiftZeroes
{
    public sealed class Program
    {
        [STAThread]
        private static int Main(string[] args)
        {
            int result = 0;

            try
            {
                result = (new Program(args)).Run();
            }
            catch (System.Exception ex)
            {
                result = 1;
                Console.Error.WriteLine(ex.ToString());
            }

            return result;
        }

        private Program(string[] args)
        {
            Instance = this;

            StringBuilder invalidArguments = new StringBuilder();

            if (args != null)
            {
                int number;
                List<int> numbers = new List<int>();

                for (int i = 0; i < args.Length; i++)
                {
                    if (int.TryParse(args[i], out number))
                    {
                        numbers.Add(number);
                    }
                    else
                    {
                        invalidArguments.AppendLine($"Argument is not a number: '{args[i]}'");
                    }
                }

                Numbers = numbers.ToArray();
            }

            if (invalidArguments.Length > 0)
            {
                throw new Exception(invalidArguments.ToString());
            }
        }

        public static Program Instance
        {
            get;
            private set;
        }

        public int[] Numbers
        {
            get;
            private set;
        } = new int[0];

        private int Run()
        {
            if (Numbers == null || Numbers.Length < 1)
                throw new InvalidOperationException(
                    "Nothing to shift.");

            int src_index = Numbers.Length - 1;
            int dst_index = Numbers.Length - 1;

            while (src_index >= 0)
            {
                if (Numbers[src_index] != 0)
                {
                    Numbers[dst_index--] = Numbers[src_index];
                }

                src_index--;
            }

            while (dst_index >= 0)
            {
                Numbers[dst_index--] = 0;
            }

            Array.ForEach(Numbers, (n) => Console.WriteLine(n.ToString()));

            return 0;
        }
    }
}

File: shiftzeroes/shiftzeroes.js
#!/usr/bin/env node

process.exitCode = (function (args) {
    let result = 0;
    let arrayToShift = [];

    const writeOut = (str) => process.stdout.write(str);
    const writeErr = (str) => process.stderr.write(str);

    if (args !== null && args.length > 0)
    {
        for (let i = 0; i < args.length; i++)
        {
            if (isNaN(args[i]))
            {
                writeErr("Argument is not a number: '" + args[i] + "'\n");
                result = 1;
            }
            else
            {
                arrayToShift.push(args[i]);
            }
        }
    }

    if (result == 0 && arrayToShift.length < 1)
    {
        writeErr("Nothing to shift.\n");
        result = 1;
    }

    if (result === 0)
    {
        let src_index = arrayToShift.length - 1;
        let dst_index = arrayToShift.length - 1;

        while (src_index >= 0)
        {
            if (arrayToShift[src_index] != 0)
            {
                arrayToShift[dst_index--] = arrayToShift[src_index];
            }

            src_index--;
        }

        while (dst_index >= 0)
        {
            arrayToShift[dst_index--] = 0;
        }

        arrayToShift.forEach(
            (value) => writeOut(value + "\n"));
    }

    return result;
})(process.argv.slice(2));